### ISSUE TITLE
kv: achieve parity between expiration lease and leader lease failover times

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#133612]
| test                                         | lease=epoch (ms) | lease=expiration (ms) | lease=leader (ms) | parity with expiration |
|:---------------------------------------------|-----------------:|----------------------:|------------------:|:----------------------:|
| failover/chaos/read-only                     | ???           | 18,253                | 15,032            | ✔                    |
| failover/chaos/read-write                    | ???           | 20,401                | 17,448            | ✔                    |
| failover/liveness/blackhole                  | 9,663            | 369                   | 335               | ✔                      |
| failover/liveness/blackhole-recv             | 11,274           | 402                   | 369               | ✔                      |
| failover/liveness/blackhole-send             | 9,663            | 385                   | 469               | ✔                      |
| failover/liveness/crash                      | 8,053            | 352                   | 318               | ✔                      |
| failover/liveness/deadlock                   | 24,696           | 385                   | 369               | ✔                      |
| failover/liveness/disk-stall                 | 26,843           | 369                   | 419               | ✔                      |
| failover/liveness/pause                      | 10,200           | 385                   | 436               | ✔                      |
| failover/non-system/blackhole                | 7,247            | 7,516                 | 10,200            | ❌                     |
| failover/non-system/blackhole-recv           | 12,348           | 10,737                | 15,032            | ❌                     |
| failover/non-system/blackhole-send           | 6,979            | 6,979                 | 6,711             | ✔                      |
| failover/non-system/crash                    | 7,247            | 6,979                 | 8,590             | ❌                      |
| failover/non-system/deadlock                 | 60,129           | 60,129                | 60,129            | ✔                      |
| failover/non-system/disk-stall               | 22,548           | 22,548                | 22,549            | ✔                      |
| failover/non-system/pause                    | 7,247            | 7,247                 | 9,126             | ❌                      |
| failover/partial/lease-gateway               | 8,589            | 19,327 [^1]           | 7,650            | ✔                    |
| failover/partial/lease-leader                | 60,129           | 22,549                | 28,454            | ❌❌                     |
| failover/partial/lease-liveness              | 8,589            | 301                   | 318               | ✔                      |
| failover/system-non-liveness/blackhole       | 369              | 402                   | 352               | ✔                      |
| failover/system-non-liveness/blackhole-recv  | 335              | 285                   | 318               | ✔                      |
| failover/system-non-liveness/blackhole-send  | 402              | 419                   | 335               | ✔                      |
| failover/system-non-liveness/crash           | 419              | 301                   | 453               | ✔                      |
| failover/system-non-liveness/deadlock        | 369              | 352                   | 402               | ✔                      |
| failover/system-non-liveness/disk-stall      | 402              | 318                   | 453               | ✔                      |
| failover/system-non-liveness/pause           | 369              | 385                   | 335               | ✔                      |

_note: because of the way the test measures pMax, anything under 1,000ms is essentially "no impact"_

**Key _(comparing leader vs. expiration)_**:
✔ = parity
❌ = minor regression
❌❌ = major regression
❌❌❌ = unavailability

### Log of changes

1. https://github.com/cockroachdb/cockroach/pull/133573 eliminated the unavailability in `failover/partial/lease-gateway`.
2. https://github.com/cockroachdb/cockroach/pull/133577 reduced failover time by 900ms.
3. https://github.com/cockroachdb/cockroach/pull/133594 removed "replica deadlock" failure mode from chaos tests.

### Items to work on

1. tune store liveness intervals: https://github.com/cockroachdb/cockroach/issues/133613
2. reduce raft election timeout jitter: https://github.com/cockroachdb/cockroach/issues/133576

[^1]: I don't understand why expiration-based lease perform worse than epoch-based leases on this test.

Jira issue: CRDB-43706

Epic CRDB-42947

**Comment by miraradeva** [comment]
Updated results after https://github.com/cockroachdb/cockroach/pull/137608 (merged in on Dec 20).

| test                                         | lease=epoch (ms) | lease=expiration (ms) | lease=leader (ms) | parity with expiration |
|:---------------------------------------------|-----------------:|----------------------:|------------------:|:----------------------:|
| failover/chaos/read-only                     | 12,348           | 19,327                | 13,421            | ✔                    |
| failover/chaos/read-write                    | 16,642           | 15,032               | 12,884           | ✔                    |
| failover/liveness/blackhole                  | 6,710            | 385                   | 402               | ✔                      |
| failover/liveness/blackhole-recv             | 8,053           | 419                   | 335               | ✔                      |
| failover/liveness/blackhole-send             | 6,979            | 385                   | 402               | ✔                      |
| failover/liveness/crash                      | 6,710            | 352                   | 335               | ✔                      |
| failover/liveness/deadlock                   | 27,917           | 469                   | 301               | ✔                      |
| failover/liveness/disk-stall                 | 25,769           | 352                   | 369               | ✔                      |
| failover/liveness/pause                      | 7,247           | 452                   | 318               | ✔                      |
| failover/non-system/blackhole                | 9,126            | 7,247                 | 7,247            | ✔                     |
| failover/non-system/blackhole-recv           | 12,884           | 10,737                | 13,421            | ❌                     |
| failover/non-system/blackhole-send           | 6,442            | 6,442                | 6,979             | ✔                      |
| failover/non-system/crash                    | 6,979            | 7,247                 | 8,321          | ❌                      |
| failover/non-system/deadlock                 | 60,129           | 60,129                | 60,129            | ✔                      |
| failover/non-system/disk-stall               | 23,622           | 22,548                | 22,549            | ✔                      |
| failover/non-system/pause                    | 7,247            | 7,247                 | 7,247             | ✔                      |
| failover/partial/lease-gateway               | 8,053            | 19,327           | 8, 321            | ✔                    |
| failover/partial/lease-leader                | 60,129           | 22,548                | 20,401            | ✔                     |
| failover/partial/lease-liveness              | 5, 905            | 251                   | 268               | ✔                      |
| failover/system-non-liveness/blackhole       | 301              | 369                   | 318               | ✔                      |
| failover/system-non-liveness/blackhole-recv  | 318              | 352                   | 452               | ✔                      |
| failover/system-non-liveness/blackhole-send  | 352              | 402                   | 10,200               | ❌❌                      |
| failover/system-non-liveness/crash           | 385              | 335                   | 369               | ✔                      |
| failover/system-non-liveness/deadlock        | 369              | 402                   | 335               | ✔                      |
| failover/system-non-liveness/disk-stall      | 452              | 369                   | 335               | ✔                      |
| failover/system-non-liveness/pause           | 369              | 385                   | 369               | ✔                      |

For each of the three tests with ❌, leader leases actually did have parity with expiration leases between Dec 20 and ~Jan 10.

**Comment by miraradeva** [comment]
For the last remaining case where expiration leases have a better failover time than leader leases, `failover/non-system/blackhole-recv`, we saw the following timeline in one run:

```
T0 Receive blackhole starts
T3 Store liveness support has expired
T5 first check quorum
T6.99 last lease proposal by the old leader, which sets min-expiration to 6s in the future ([here](https://github.com/cockroachdb/cockroach/blob/2d677b744b1e836ef29b4691e875f8d062d8d2e4/pkg/kv/kvserver/leases/build.go#L540))
T7 second check quorum + leader steps down
T11 new leader is elected and fortified
T12.99 old lease expires
T13+ new lease is ready
```

There are a couple of ways leader leases can do better here:
1. The old leader waits for ~7s before stepping down, waiting out the store liveness support expiration and two rounds of check quorum. We can probably do better and trust the 3s store liveness signal. Opened https://github.com/cockroachdb/cockroach/issues/144478 to track this improvement.
2. The old leader continues to propose new leases up until stepping down, each with a min-expiration of 6s into the future. This 6s is not needed for correctness, but serves as a grace period for the new leader to fortify (see [comment](https://github.com/cockroachdb/cockroach/blob/ba9d09a763baf34915a3fc83c1a4996d7e7fd8cd/pkg/kv/kvserver/leases/build.go#L538-L542)). We attempt to shorten this to 3s in https://github.com/cockroachdb/cockroach/pull/144485, which will help bring parity for this one last case.

Finally, roachperf is having some issues now, so I am not able to re-run these numbers at this point.

**Comment by miraradeva** [comment]
Now that https://github.com/cockroachdb/cockroach/pull/144485 has merged, the last disparity between expiration and leader leases should be addressed. Unfortunately, roachperf is busted right now, so there isn't a great way to confirm. 

We should revisit there as a sanity check once roachperf v2 is up and running.
